doc/neps/nep-0050-scalar-promotion.rst

```diff
@@ -0,0 +1,729 @@
+more strict:  This means that the results will be unchanged in the first
+two examples.  For the second one, the results will be the following::
+
+  np.arange(10, dtype=np.int8) + 256  # raises a TypeError
+  np.array([1., 2.], dtype=np.float32) * 1e200  # warning and returns infinity
+
+The second one will warn because ``np.float32(1e200)`` overflows to infinity.
+It will then do the calculation with ``inf`` as normally.
+
+
+.. admonition:: Behaviour in other libraries
+
+   Overflowing in the conversion rather than raising an error is a choice;
+   it is one that is the default in most C setups (similar to NumPy C can be
+   set up to raise an error due to the overflow, however).
+   It is also for example the behaviour of ``pytorch`` 1.10.
+
+
+
+Backward compatibility
+======================
+
+In general, code which only uses the default dtypes float64, or int32/int64
+or more precise ones should not be affected.
+
+However, the proposed changes will modify results in quite a few cases where
+0-D or scalar values (with non-default dtypes) are mixed.
+In many cases, these will be bug-fixes, however, there are certain changes
+which may be particularly interesting.
+
+The most important failure is probably the following example::
+
+  arr = np.arange(100, dtype=np.uint8)  # storage array with low precision
+  value = arr[10]
+
+  # calculation continues with "value" without considering where it came from
+  value * 100
+
+Where previously the ``value * 100`` would cause an up-cast to int32/int64
+(because value is a scalar).  The new behaviour will preserve the lower
+precision unless explicitly dealt with (just as if ``value`` was an array).
+This can lead to integer overflows and thus incorrect results beyond precision.
+In many cases this may be silent, although NumPy usually gives warnings for the
+scalar operators.
+
+Similarliy, if the storage array is float32 a calculation may retain the lower
+float32 precision rather than use the default float64.
+
+Further issues can occur.  For example:
+
+* Floating point comparisons, especially equality, may change when mixing
+  precisions::
+
+     np.float32(1/3) == 1/3  # was False, will be True.
+
+* Certain operations are expected to start failing::
+
+     np.array([1], np.uint8) * 1000
+     np.array([1], np.uint8) == 1000  # possibly also
+
+  to protect users in cases where previous value-based casting led to an
+  upcast.  (Failures occur when converting ``1000`` to a ``uint8``.)
+
+* Floating point overflow may occur in odder cases::
+
+     np.float32(1e-30) * 1e50  # will return ``inf`` and a warning
+
+  Because ``np.float32(1e50)`` returns ``inf``.  Previously, this would return
+  a double precision result even if the ``1e50`` was not a 0-D array
+
+In other cases, increased precision may occur.  For example::
+
+  np.multiple(float32_arr, 2.)
+  float32_arr * np.float64(2.)
+
+Will both return a float64 rather than float32.  This improves precision but
```

Comment:
I thought python scalars would be treated as "weak"? Why does the first one upcast to float64?

Reply:
I think I use "Python scalars" for `int`, `float`, `complex`.  So the ones builtin with Python, not the *NumPy* scalars.  But maybe I can try to avoid that term to avoid the tricky distinction.